src/core/geometry/qgspolygon.cpp

```diff
@@ -185,6 +185,50 @@ QByteArray QgsPolygon::asWkb( QgsAbstractGeometry::WkbFlags flags ) const
   return wkbArray;
 }
 
+QString QgsPolygon::asWkt( int precision ) const
```

Comment:
It seems there might be confusion in handling the geometry types. Your test highlights the degenerate WKT output resulting from adding a curve to your polygon. The output generated is `Polygon ((0 0, 0 10, 10 10, 10 0, 0 0),CompoundCurve ((1 1, 1 9, 9 9, 9 1, 1 1)))`, which is invalid. To rectify this, it should ideally be of the type `CurvePolygon ((0 0, 0 10, 10 10, 10 0, 0 0),CompoundCurve ((1 1, 1 9, 9 9, 9 1, 1 1)))`, but that belongs to `QgsCurvePolygon `and not `QgsPolygon`.

To ensure incoming geometries into `QgsPolygon `are segmented properly, it's essential to verify that the layer is indeed of type Polygon and not CurvePolygon. 

As it stands, this fixes the issue, but I'm not sure if it might be masking the real problem that has been occurring in the recent editing versions. AFAIR, it worked before (c)

Reply:
I agree, and we already discussed that in https://github.com/qgis/QGIS/pull/55588#discussion_r1425165009

I don't really know when it was broken, my gut feeling is that it was broken when the maptool started generating compound curves (even when not necessary). 

But, as I already wrote, the real issue IMO is that the API allows to construct an "invalid" QgsPolygon, but we cannot change that without breaking the API. 

So, I cannot think of another solution than to be good citizens and to produce valid WKT from and "invalid" polygon.